### PR TITLE
chore: setup expo preview

### PR DESCRIPTION
**Pull request by allforabit**

### CODE REVIEW

**Comment by allforabit** [comment]
@simoneb there seems to be some issues with the expo action and the QR code is incorrect. See issue here: https://github.com/expo/expo-preview-action/issues/6 I can find the project link in the command line output and that points here: https://expo.dev/@nearform/optic-expo?release-channel=pr-246 However that doesn't seem to be working on my device.

**Comment by simoneb** [comment]
Before we troubleshoot this can you please align it to how it's done in https://github.com/nearform/optic-expo/blob/master/.github/workflows/publish.yml? E.g. the environment handling

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -24,3 +24,11 @@ jobs:
...
       - run: yarn typecheck
       - run: yarn test
+
+  deploy-expo-preview:
+    name: deploy expo preview
+    if: github.event_name == 'pull_request'
+    needs: [build]
+    uses: nearform/optic-expo/.github/workflows/deploy-expo-preview.yml@chore/expo-preview
+    secrets:
+      expo-token: ${{ secrets.EXPO_TOKEN }}
```
Comment (simoneb) [inline .github/workflows/ci.yml:34]: nit: add new line to avoid this red mark

**Review by simoneb — COMMENTED** [review]
This works but I think we can do better. There is quite a bit of duplication between the existing publish workflow and this new reusable workflow.

They are doing exactly the same thing except for the handling of the release channel. I would suggest that:

1. we create a single reusable workflow which does the publishing
2. it takes as **input** the release channel, with the default value being default
3. it's fine to always comment on the PR even when the release channel is default
4. invoke this reusable workflow both from the publish workflow and from the ci workflow. In the latter case, it will provide a release-channel input with value `pr-${{ github.event.number }}`

Minor:

- I prefer removing all the names on the steps, they're usually redundant and unnecessarily clutter the workflow code
- point to specific versions of third party actions instead of master

**Comment by allforabit** [comment]
I think it's right now and hope it's what you had in mind as far as sharing the expo publishing action. One thing is it's set to use this current branch ( `uses: nearform/optic-expo/.github/workflows/expo-publish.yml@chore/expo-preview`) but should be master. The build will fail if I do this though.

File: .github/workflows/cd.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+name: publish
```
Comment (simoneb) [inline .github/workflows/cd.yml:1]: let's keep the name consistent with the file name, hence: cd

File: .github/workflows/cd.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  push:
+    branches:
+      - master
+
+jobs:
+  publish:
+    name: publish to expo
+    uses: nearform/optic-expo/.github/workflows/expo-publish.yml@chore/expo-preview
+    with:
+      channel: default
```
Comment (simoneb) [inline .github/workflows/cd.yml:13]: minor: shall we make this the default in the reusable workflow so if it's not specified it goes there?

File: .github/workflows/expo-publish.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+jobs:
+  deploy-expo-preview:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up repository
+        uses: actions/checkout@v2
+      - name: Set up Node
+        uses: actions/setup-node@v1
+        with:
+          node-version: 14.x
```
Comment (simoneb) [inline .github/workflows/expo-publish.yml:23]:
```suggestion
          node-version: 14
```

File: .github/workflows/expo-publish.yml [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        type: string
+
+jobs:
+  deploy-expo-preview:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up repository
+        uses: actions/checkout@v2
+      - name: Set up Node
+        uses: actions/setup-node@v1
```
Comment (simoneb) [inline .github/workflows/expo-publish.yml:21]: please remove names. use latest versions of these actions

File: .github/workflows/expo-publish.yml [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        with:
+          node-version: 14.x
+      - name: Set up Expo
```
Comment (simoneb) [inline .github/workflows/expo-publish.yml:24]: remove name

File: .github/workflows/expo-publish.yml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        run: cp .env.sample .env
+      - name: Publish to Expo
+        run: expo publish --release-channel=${{ inputs.channel }}
+      - name: Pr Comment deployment link
+        if: github.event_name == 'pull_request'
+        uses: unsplash/comment-on-pr@master
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          msg: App is ready for review, you can [see it here](https://expo.dev/@nearform/optic-expo?release-channel=pr-${{ github.event.number }}).
```
Comment (simoneb) [inline .github/workflows/expo-publish.yml:42]: remove all names, they're unnecessary clutter

File: .github/workflows/cd.yml (hunk 2 above)
Comment (simoneb) [inline .github/workflows/cd.yml:1]: since you're removing the publish workflow, change the readme to include this one please

**Review by simoneb — COMMENTED** [review]
Some comments but this is the implementation I had in mind 👌

**Comment by allforabit** [comment]
I've set this to run on the action from the master branch so it's failing. It'll work once it's merged in.

File: .github/workflows/expo-publish.yml [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+          expo-token: ${{ secrets.expo-token }}
+      - run: yarn install
+      - run: cp .env.sample .env
+      - run: expo publish --release-channel=${{ inputs.channel }}
+      - uses: unsplash/comment-on-pr@master
+        if: github.event_name == 'pull_request'
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          msg: App is ready for review, you can [see it here](https://expo.dev/@nearform/optic-expo?release-channel=pr-${{ github.event.number }}).
```
Comment (simoneb) [inline .github/workflows/expo-publish.yml:36]:
```suggestion
          msg: App is ready for review, you can [see it here](https://expo.dev/@nearform/optic-expo?release-channel=${{ inputs.channel }}).
```

File: README.md [hunk 9]
```diff
@@ -1,7 +1,7 @@
 # Optic Expo
 
 [![ci](https://github.com/nearform/optic-expo/actions/workflows/ci.yml/badge.svg)](https://github.com/nearform/optic-expo/actions/workflows/ci.yml)
-[![publish](https://github.com/nearform/optic-expo/actions/workflows/publish.yml/badge.svg)](https://github.com/nearform/optic-expo/actions/workflows/publish.yml)
+[![publish](https://github.com/nearform/optic-expo/actions/workflows/cd.yml/badge.svg)](https://github.com/nearform/optic-expo/actions/workflows/publish.yml)
```
Comment (simoneb) [inline README.md:4]:
```suggestion
[![cd](https://github.com/nearform/optic-expo/actions/workflows/cd.yml/badge.svg)](https://github.com/nearform/optic-expo/actions/workflows/cd.yml)
```

**Review by simoneb — CHANGES_REQUESTED** [review]

**Review by simoneb — APPROVED** [review]

**Pull request merged**